README.md
# Changelog Manager
A small command line utility to manage changelogs and automate generating CHANGELOG.md

## Overview

The compiled Changelog Manager executable can just be dropped into any development project, like a `[project root]/bin` or `[project root]/tools` directory.  It can be used to generate json files representing a change log entry and used to combine all the files into your project's CHANGELOG.md as well as generate a standalone [build]-CHANGELOG.md in the project root directory.

> note: Changelog Manger currently uses `git rev-parse --show-toplevel` to set `[project root]`

## Usage

Changelog Manager has four commands:
 * `version` - gets the current changelogmanager version
 * `help` - displays the terminal help for changelogmanager
 * `add` - generates changelog entry .json files
 * `combine` - combines the existing changelog entry files and outputs two files
> note: all commands support -h or --help thanks to [Cobra](https://github.com/spf13/cobra)
 
### add
**add** is used to output generated .json files to `[project root]/changelogs/unreleased/`.  This command has four flags that can be set when running executing it.
* Required flags:
    * `--description [string]` or `-d [string]` - the main message that will be displayed in the changelog for the corresponding code changes.
    * `--author [string]` or `-a [string]` - name of the developer responsible for the changes
* Optional flags:
    * `--merge [id]` or `-m [id]` - sets the merge or pull request id associated with the changes
    * `--platform [string]` or `-p [string]` - logs the platform the changes were applied to
    
#### Example Usage
`changelogmanager add -d "This is my change description" -a "John Smith" -m 123`

### combine
**combine** takes all the .json file in `[project root]/changelogs/unreleased/`, combines them into a chunk of markdown, then appends the resulting markdown to the top of your `[project root]/CHANGELOG.md` as well as generating a `[project root]/[build]-CHANGELOG.md` stand alone file.  By default it will delete the .json files after updating CHANGELOG.md. Combine only has one required flag:
* Required flags:
    * `--build [string]` or `-b [string]` - sets the build/version number all the changes are going to be applied to.
* Optional flags:
    * `--note [string]` or `-n [string]` - sets a note for the build
    * `--keep` or `-k` - retains the .json files after the combine is complete
    
#### Example Usage
`changelogmanager combine -b "v0.1" -note "first build hosted on gitlab" -k`


CHANGELOG.md
## v0.2 (2017-03-23)
**Mac and Windows builds uploaded to github**
* Adds archive flag to combine to save json files - *gspeager* - **Merge:** 6
* Adds support for different markdown flavors - *gspeager* - **Merge:** 4
* Adds user flag to add command to use git username instead of author - *gspeager* -  **Merge:** 5



add.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"strings"
	"time"

	"github.com/pkg/errors"
	"github.com/spf13/cobra"
)

var (
	author   string
	merge    string
	message  string
	platform string
	entry    ChangelogEntry
	user     bool
)

// addCmd represents the add command
var addCmd = &cobra.Command{
	Use:   "add",
	Short: "Adds a new changelog entry file",
	Long: `Adds a changelog entry file requiring description and author. Example usage:
	
	add -a="John Smith" -d="full description text here" -m="123"`,
	Run: func(cmd *cobra.Command, args []string) {

		if len(author) == 0 && !user {
			fmt.Println("Either author or user is required see --help for usage")
			return
		} else if len(message) == 0 {
			fmt.Println("Description is required see --help for usage")
			return
		}
		authorName := author
		if user {
			authorName, _ = getGitUsername()
		}

		entry.Author = authorName
		entry.Platform = strings.ToLower(platform)
		entry.Message = message
		entry.Merge = merge

		entryJSON, err := json.Marshal(entry)
		if err != nil {
			fmt.Println(err)
			return
		}
		if _, err := os.Stat(changelogEntriesPath); os.IsNotExist(err) {
			pathErr := os.MkdirAll(changelogEntriesPath, 0777)
			if pathErr != nil {
				fmt.Println(err)
				return
			}
		}
		filename := strings.Replace(strings.Fields(time.Now().String())[0], "-", "", -1)
		filename += "-" + sanitizeDescription(entry.Message) + ".json"
		err = ioutil.WriteFile(changelogEntriesPath+filename, entryJSON, 0644)
		if err != nil {
			fmt.Println(err)
		}
	},
}

func init() {
	RootCmd.AddCommand(addCmd)
	addCmd.Flags().StringVarP(&author, "author", "a", "", "Required: name of the user submitting the message")
	addCmd.Flags().StringVarP(&message, "description", "d", "", "Required: description to summarize changes")
	addCmd.Flags().StringVarP(&merge, "merge", "m", "", "Merge request id")
	addCmd.Flags().StringVarP(&platform, "platform", "p", "", "Platform affected by changes")
	addCmd.Flags().BoolVarP(&user, "user", "u", false, "Use instead of author to pull the git username will override author")
}

func sanitizeDescription(msg string) string {
	if msg == "" {
		return msg
	}
	result := msg
	invalidChars := []string{"*", "`", "_", "~", "]", "[", "!", "#", "(", ")"}

	for _, c := range invalidChars {
		result = strings.Replace(result, c, "", -1)
	}

	result = strings.Replace(result, " ", "-", -1)

	return result
}

func getGitUsername() (string, error) {
	gitCmd := exec.Command("git", "config", "user.name")

	b := &bytes.Buffer{}
	gitCmd.Stdout = b

	err := gitCmd.Run()
	if err != nil {
		return "", errors.Wrap(err, "cannot get git user name")
	}

	output := b.String()

	if len(output) == 0 {
		return "", errors.New("git username had no output")
	}

	return strings.TrimSpace(output), nil
}


main.go
package main

import (
	"bytes"
	"os/exec"
	"strings"

	"github.com/pkg/errors"
)

var (
	rootPath             string
	changelogEntriesPath string
	changelogArchivePath string
	changelogFilePath    string
)

const version = "0.1"

// ChangelogEntry as a databucket for log entry data
type ChangelogEntry struct {
	Message  string `json:"message"`
	Author   string `json:"author"`
	Platform string `json:"platform"`
	Merge    string `json:"merge"`
}

func main() {
	path, _ := getAppPath()
	rootPath = path
	changelogEntriesPath = path + "changelogs/unreleased/"
	changelogArchivePath = path + "changelogs/released/"
	changelogFilePath = path + "CHANGELOG.md"

	Execute()
}

func getAppPath() (string, error) {
	gitCmd := exec.Command("git", "rev-parse", "--show-toplevel")

	b := &bytes.Buffer{}
	gitCmd.Stdout = b

	err := gitCmd.Run()
	if err != nil {
		return "", errors.Wrap(err, "cannot find app root dir git rev-parse failed")
	}

	output := b.String()

	if len(output) == 0 {
		return "", errors.New("cannot find app root dir git rev-parse had no output")
	}

	return strings.TrimSpace(output) + "/", nil
}


root.go
package main

import (
	"fmt"
	"os"

	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

var cfgFile string

// RootCmd represents the base command when called without any subcommands
var RootCmd = &cobra.Command{
	Use:   "changelogmanager",
	Short: "Small app to manage change log entries and combine them into CHANGELOG.md",
	Long: `Allows users to generate change log entry .json files /changelogs/unreleased/.
	
	Then can combine all the .json files in /changelogs/unreleased/ 
	into CHANGELOG.md and a stand alone file`,
}

// Execute adds all child commands to the root command sets flags appropriately.
// This is called by main.main(). It only needs to happen once to the rootCmd.
func Execute() {
	if err := RootCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(-1)
	}
}

func init() {
	cobra.OnInitialize(initConfig)
}

// initConfig reads in config file and ENV variables if set.
func initConfig() {
	if cfgFile != "" { // enable ability to specify config file via flag
		viper.SetConfigFile(cfgFile)
	}

	viper.SetConfigName(".changelogmanager") // name of config file (without extension)
	viper.AddConfigPath("$HOME")             // adding home directory as first search path
	viper.AutomaticEnv()                     // read in environment variables that match

	// If a config file is found, read it in.
	if err := viper.ReadInConfig(); err == nil {
		fmt.Println("Using config file:", viper.ConfigFileUsed())
	}
}


formatters.go
package main

import (
	"bytes"
	"fmt"
	"strings"
	"time"
)

func buildFlavoredContent(flavor string, b *bytes.Buffer, changelogEntries []ChangelogEntry) {
	mergeIdentifier := "#"
	if flavor == "gitlab" {
		mergeIdentifier = "!"
	}
	header := build + " (" + strings.Fields(time.Now().String())[0] + ")"
	fmt.Fprintf(b, "## %s\n", header)
	if note != "" {
		fmt.Fprintf(b, "**%s**\n", note)
	}
	for _, e := range changelogEntries {
		result := "* "

		if e.Platform != "" {
			result += "[" + e.Platform + "] "
		}
		result += e.Message + " @" + e.Author

		if e.Merge != "" {
			result += " " + mergeIdentifier + e.Merge + "\n"
		}
		fmt.Fprint(b, result)
	}

	fmt.Fprintf(b, "\n")
}

func unflavoredContent(b *bytes.Buffer, changelogEntries []ChangelogEntry) {
	header := build + " (" + strings.Fields(time.Now().String())[0] + ")"
	fmt.Fprintf(b, "## %s\n", header)
	if note != "" {
		fmt.Fprintf(b, "**%s**\n", note)
	}
	for _, e := range changelogEntries {
		result := "* "

		if e.Platform != "" {
			result += "[" + e.Platform + "] "
		}
		result += e.Message + " - *" + e.Author + "*"

		if e.Merge != "" {
			result += " - **Merge:** " + e.Merge + "\n"
		}
		fmt.Fprint(b, result)
	}

	fmt.Fprintf(b, "\n")
}


combine.go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"

	"encoding/json"

	"github.com/spf13/cobra"
)

var (
	build   string
	keep    bool
	note    string
	flavor  string
	archive bool
)

// combineCmd represents the combine command
var combineCmd = &cobra.Command{
	Use:   "combine",
	Short: "Combines existing changelog entries into CHANGELOG.md",
	Long: `combines all the .json files in the /changelog/unreleased/ folder
	and appends to a master CHANGELOG.md 
	as well as creating a [build]-CHANGELOG.md. Example usage:
	
	combine -b "v0.1" -n "released to app store"`,
	Run: func(cmd *cobra.Command, args []string) {
		if _, err := os.Stat(changelogEntriesPath); os.IsNotExist(err) {
			fmt.Println(err)
			return
		}

		if _, err := os.Stat(changelogArchivePath); os.IsNotExist(err) {
			pathErr := os.MkdirAll(changelogArchivePath, 0777)
			if pathErr != nil {
				fmt.Println(err)
				return
			}
		}

		if len(build) == 0 {
			fmt.Println("Build is required see --help for usage")
			return
		}
		if archive {
			if _, err := os.Stat(changelogArchivePath + sanitizeDescription(build) + "/"); os.IsNotExist(err) {
				pathErr := os.MkdirAll(changelogArchivePath+sanitizeDescription(build)+"/", 0777)
				if pathErr != nil {
					fmt.Println(err)
					return
				}
			}
		}

		//Parse files
		var changelogFiles []os.FileInfo
		var changelogEntries []ChangelogEntry

		files, err := ioutil.ReadDir(changelogEntriesPath)
		if err != nil {
			fmt.Println(err)
			return
		}

		for _, f := range files {
			if filepath.Ext(f.Name()) == ".json" {
				changelogFiles = append(changelogFiles, f)

				contents, err := ioutil.ReadFile(changelogEntriesPath + f.Name())
				if err != nil {
					fmt.Println(err)
					return
				}

				//archive old changelogFiles
				if archive {
					err = ioutil.WriteFile(changelogArchivePath+sanitizeDescription(build)+"/"+f.Name(), contents, 0644)
					if err != nil {
						fmt.Println(err)
					}
				}

				var entry ChangelogEntry
				err = json.Unmarshal(contents, &entry)
				if err != nil {
					fmt.Println(err)
					return
				}

				changelogEntries = append(changelogEntries, entry)
			}
		}

		//Build output
		b := &bytes.Buffer{}
		if len(flavor) == 0 {
			unflavoredContent(b, changelogEntries)
		} else {
			buildFlavoredContent(strings.ToLower(flavor), b, changelogEntries)
		}

		//Write files
		os.OpenFile(changelogFilePath, os.O_RDONLY|os.O_CREATE, 0666)

		fileContent, err := ioutil.ReadFile(changelogFilePath)
		if err != nil {
			fmt.Println(err)
			return
		}

		ioutil.WriteFile(changelogFilePath, b.Bytes(), 0644)

		f, err := os.OpenFile(changelogFilePath, os.O_APPEND|os.O_WRONLY, 0600)
		if err != nil {
			fmt.Println(err)
			return
		}

		defer f.Close()

		if _, err = f.Write(fileContent); err != nil {
			fmt.Println(err)
			return
		}

		ioutil.WriteFile(changelogArchivePath+build+"-CHANGELOG.md", b.Bytes(), 0644)

		//remove old changelogFiles
		if !keep {
			for _, f := range changelogFiles {
				os.Remove(changelogEntriesPath + f.Name())
			}
		}
	},
}

func init() {
	RootCmd.AddCommand(combineCmd)
	combineCmd.Flags().StringVarP(&build, "build", "b", "", "Required: Build version used as header in .md file")
	combineCmd.Flags().StringVarP(&note, "note", "n", "", "Note about this build")
	combineCmd.Flags().BoolVarP(&keep, "keep", "k", false, "Maintains the .json files in the unreleased directory")
	combineCmd.Flags().StringVarP(&flavor, "flavor", "f", "", "Sets markdown flavor, can be 'github' or 'gitlab'")
	combineCmd.Flags().BoolVarP(&archive, "archive", "a", false, "Archives the .json files to changelogs/released/[build]/")
}
